### ISSUE TITLE
How to train SCNN on my own dataset

### ISSUE THREAD

**Issue by danyan0820** [XingangPan/SCNN#7]
the imagesize of my own dataset is 1080*1920,what should I do if I want to train SCNN on this dataset

**Comment by XingangPan** [comment]
Firstly, you should modify data augmentation parameters to yours in function "preprocess_aug()" at datasets/lane.lua. (Note: the final image height and weight should be multiple of 16, like 368 * 640)
Secondly, modify the pretrained VGG model, change the parameters of nn.View and nn.Linear layer of the classification branch to yours. You can do so by remove the original layers and insert new layers. For example, if your input image size is 368*640, then the paramter "4500" should be changed to "4600" (5*368*640/16/16).
Thirdly, modify experiments/models/SCNN-gen.lua line 11-15. Set value of "Num" and "length" to yours. For example, if your input image size is 368*640, then "36" should be changed to "46", "100" should be changed to "80". Then you should be able to generate the right SCNN model.

**Comment by XingangPan** [comment]
Firstly, you should modify data augmentation parameters to yours in function "preprocess_aug()" at datasets/lane.lua. (Note: the final image height and weight should be multiple of 16, like 368x640)
Secondly, modify the pretrained VGG model, change the parameters of nn.View and nn.Linear layer of the classification branch to yours. You can do so by remove the original layers and insert new layers. For example, if your input image size is 368x640, then the paramter "4500" should be changed to "4600" (5x368x640/16/16).
Thirdly, modify experiments/models/SCNN-gen.lua line 11-15. Set value of "Num" and "length" to yours. For example, if your input image size is 368x640, then "36" should be changed to "46", "100" should be changed to "80". Then you should be able to generate the right SCNN model.

**Comment by juntawu** [comment]
@danyan0820 Have you trained SCNN with your own dataset? I am stuck with it! Can you please give me some guide ?? Thanks a lot!

**Comment by juntawu** [comment]
@XingangPan Can you please give more detailed information about this issue? Thanks a lot !!

**Comment by XingangPan** [comment]
@cuicuizhang1989 @juntawu For example, in order to change the nn.View(4500) to nn.View(4600), you can do as follows:
require 'nn'
require 'cudnn'
model = torch.load('vgg/vgg.t7')
classifier = model:get(44):get(2)
classifier:remove(3)
classifier:insert(nn.View(4600), 3)
print(model)

In a word, the 'get', 'remove', and 'insert' function of the 'Containers' class will do. Check https://github.com/torch/nn/blob/master/doc/containers.md#nn.Containers for more details.

**Comment by juntawu** [comment]
@XingangPan  I want to train images with size 640 * 512. As you guided, i modified function "preprocess_aug()", "SCNN-gen.lua" and "nn.View". However i got this error during training

`DataParallelTable.lua:617: bad argument #1 to 'copy' (sizes do not match at /home/lishanliao/torch/extra/cutorch/lib/THC/THCTensorCopy.cu:31)`

Is there anything else to change?

**Comment by YgRen** [comment]
@XingangPan so,I have a question, how to fit your model to my own dataset which is 0~18 lane lines.  and it is not  parallel four lanes.can you tell me how  to prepare the dataset for this model? so that  it can be trained. thanx!

**Comment by cardwing** [comment]
@YgRen , you can just modify the output of probability maps branch from 5 x 288 x 800 to 20 x 288 x 800 (the first dimension is channel, the other two are height and width, respectively) and the output of lane existence branch from 1 x 4 to 1 x 19.

**Comment by virajdeokar** [comment]
@cardwing Which one is the lane existence branch? Is it the model:get(44):get(2):get(6), the nn.linear(128-->4)?

**Comment by passion3394** [comment]
@cardwing I use the SCNN-gen.lua file to generate my own network. But I got the following error, I have find everywhere on the search engine or GitHub, but no answer given, could you help me to solve it, thanks ahead!!!

/usr/local/bin/luajit: /usr/local/share/lua/5.1/nn/THNN.lua:110: weight tensor should be defined either for all or no classes at /tmp/luarocks_cunn-scm-1-1602/cunn/lib/THCUNN/generic/SpatialClassNLLCriterion.cu:28
stack traceback:
        [C]: in function 'v'
        /usr/local/share/lua/5.1/nn/THNN.lua:110: in function 'SpatialClassNLLCriterion_updateOutput'
        /usr/local/share/lua/5.1/nn/SpatialClassNLLCriterion.lua:43: in function 'updateOutput'
        ...cal/share/lua/5.1/cudnn/SpatialCrossEntropyCriterion.lua:37: in function 'updateOutput'
        ./ParallelCriterion2.lua:28: in function 'forward'
        ./train.lua:71: in function 'train'
        main.lua:51: in main chunk
        [C]: in function 'dofile'
        /usr/local/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
        [C]: at 0x00405f70

**Comment by cardwing** [comment]
@passion3394 , it seems that the size of the weight tensor defined for spatial cross entropy loss does not equal to the number of the number of output channels. Please have a look at the number of the output channels of the lane detection branch (semantic segmentation).

**Comment by cardwing** [comment]
@virajdeokar , nn.linear(128-->4) is the last layer of the lane existence prediction branch.

**Comment by passion3394** [comment]
> @passion3394 , it seems that the size of the weight tensor defined for spatial cross entropy loss does not equal to the number of output channels. Please have a look at the number of the output channels of the lane detection branch (semantic segmentation).

@cardwing thanks very much, I have solved the problem. It's because I forget to modify the lane numbers in lane-gen.lua.

**Comment by yujincheng08** [comment]
@juntawu I have exact problem:
> DataParallelTable.lua:617: bad argument #1 to 'copy' (sizes do not match at /root/torch/extra/cutorch/lib/THC/THCTensorCopy.cu:31)

And I found that in `experiement/train.sh`, the batchSize should be divided by nGPU. I used 8 GPUs and batchSize is 12 before, so the problem raised. If I changed the batchSize to 24, everything worked!

**Comment by norainnosun** [comment]
> @XingangPan so,I have a question, how to fit your model to my own dataset which is 0~18 lane lines. and it is not parallel four lanes.can you tell me how to prepare the dataset for this model? so that it can be trained. thanx!

@fimo723 Do you know how to prepare the data now?Should Id be assigned 0~18 from left to right?thanks
